### ISSUE TITLE
workers > 0 on DataLoader makes training unreasonably slow

### ISSUE THREAD

**Issue by Tripparsugo** [Lightning-AI/lightning#17410]
### Bug description

Setting workers > 0 on a DataLoader slows down training by a factor of 50-100 on cpu

### What version are you seeing the problem on?

lightning==2.0.0
lightning-cloud==0.5.32
lightning-utilities==0.8.0

### How to reproduce the bug

I used the following script to benchmark the training speed. It's a very simple model hence I would expect the fit to be almost immediate. The script takes ~150s to run with workers.

```python
import random
import time

import lightning.pytorch as pl
import torch
from torch.utils.data import DataLoader
from lightning.pytorch.loggers import TensorBoardLogger
from lightning.pytorch.callbacks import TQDMProgressBar, EarlyStopping


class Model(pl.LightningModule):
    def __init__(self):
        super(Model, self).__init__()
        self.linear1 = torch.nn.Linear(1, 10)
        self.f1 = torch.nn.ELU(inplace=False)
        self.linear2 = torch.nn.Linear(10, 1)
        self.f2 = torch.nn.ELU(inplace=False)

    def forward(self, x):
        x = self.linear1(x)
        x = self.f1(x)
        x = self.linear2(x)
        x = self.f2(x)
        return x

    def training_step(self, train_batch, batch_idx):
        # train_batch = (torch.rand(64, 324), torch.rand(64, 22))
        mse_l = torch.nn.MSELoss()
        x, y = train_batch
        logits = self.forward(x)
        loss = mse_l(logits, y)
        self.log('train_loss', loss)
        return loss

    def validation_step(self, val_batch, batch_idx):
        mse_l = torch.nn.MSELoss()
        x, y = val_batch
        logits = self.forward(x)
        loss = mse_l(logits, y)
        self.log('val_loss', loss)
        return loss

    def configure_optimizers(self):
        optimizer = torch.optim.Adam(self.parameters(), lr=1e-3)
        return optimizer


def f(x):
    return 1 + x[0] * 2 + x[0] * x[0]


if __name__ == '__main__':
    if torch.backends.mps.is_available():
        mps_device = torch.device("mps")
        x = torch.ones(1, device=mps_device)
        print(x)
    else:
        print("MPS device not found.")

    size = 1000
    xs = [[random.random() * 10] for x in range(0, size)]
    ys = [[f(x)] for x in xs]
    train_split = int(1000 * 0.8)
    validation_split = int(1000 * 0.9)

    data = list(zip([torch.tensor(x) for x in xs], [torch.tensor(y) for y in ys]))

    train_data = data[:train_split]
    validation_data = data[train_split:validation_split]
    test_data = data[validation_split:]

    train_loader = DataLoader(train_data, batch_size=10, num_workers=10)
    validation_loader = DataLoader(validation_data, batch_size=10, num_workers=10)

    # train
    trainer = pl.Trainer(max_epochs=5,
                         callbacks=[EarlyStopping(monitor="val_loss", mode="min", patience=3),
                                    TQDMProgressBar(refresh_rate=1)
                                    ],
                         log_every_n_steps=10
                         )

    model = Model()
    loss_fn = torch.nn.MSELoss()
    xs_test = torch.stack(list(list(zip(*test_data))[0]))
    ys_test = torch.stack(list(list(zip(*test_data))[1]))

    pred_bt = model(xs_test)
    loss_bt = loss_fn(pred_bt, ys_test)
    start_time = time.time()
    trainer.fit(model=model, train_dataloaders=train_loader, val_dataloaders=validation_loader)
    end_time = time.time()
    pred_at = model(xs_test)

    loss_at = loss_fn(pred_at, ys_test)

    print(f"loss before training = {loss_bt}")
    print(f"loss after training = {loss_at}")
    print(f"training took = {end_time -start_time}s")

```

The script takes ~150s to run with workers. By setting the workers to 0 it now takes about 1s.

```
train_loader = DataLoader(train_data, batch_size=10, num_workers=0)
validation_loader = DataLoader(validation_data, batch_size=10, num_workers=0)
```
```




### Environment

<details>
  <summary>Current environment</summary>

* CUDA:
        - GPU:               None
        - available:         False
        - version:           None
* Lightning:
        - lightning:         2.0.0
        - lightning-cloud:   0.5.32
        - lightning-utilities: 0.8.0
        - pytorch-lightning: 2.0.0
        - torch:             2.0.0
        - torchmetrics:      0.11.4
        - torchvision:       0.15.1
* Packages:
        - about-time:        4.2.1
        - absl-py:           1.4.0
        - aiohttp:           3.8.4
        - aiosignal:         1.3.1
        - alive-progress:    3.1.1
        - anyio:             3.6.2
        - appnope:           0.1.3
        - argon2-cffi:       21.3.0
        - argon2-cffi-bindings: 21.2.0
        - arrow:             1.2.3
        - asttokens:         2.2.1
        - astunparse:        1.6.3
        - async-timeout:     4.0.2
        - attrs:             22.2.0
        - backcall:          0.2.0
        - beautifulsoup4:    4.12.0
        - bleach:            6.0.0
        - blessed:           1.20.0
        - cachetools:        5.3.0
        - certifi:           2022.12.7
        - cffi:              1.15.1
        - charset-normalizer: 3.1.0
        - click:             8.1.3
        - cloudpickle:       2.2.1
        - comm:              0.1.3
        - contourpy:         1.0.7
        - croniter:          1.3.8
        - cycler:            0.11.0
        - dateutils:         0.6.12
        - debugpy:           1.6.6
        - decorator:         5.1.1
        - deepdiff:          6.3.0
        - defusedxml:        0.7.1
        - dnspython:         2.3.0
        - elosports:         0.1.1
        - email-validator:   1.3.1
        - etils:             1.1.0
        - executing:         1.2.0
        - fastapi:           0.88.0
        - fastjsonschema:    2.16.3
        - filelock:          3.10.6
        - flatbuffers:       23.3.3
        - fonttools:         4.39.3
        - fqdn:              1.5.1
        - frozenlist:        1.3.3
        - fsspec:            2023.3.0
        - gast:              0.4.0
        - google-auth:       2.16.2
        - google-auth-oauthlib: 0.4.6
        - google-pasta:      0.2.0
        - grapheme:          0.6.0
        - grpcio:            1.51.3
        - gym:               0.26.2
        - gym-notices:       0.0.8
        - h11:               0.14.0
        - h5py:              3.8.0
        - httpcore:          0.16.3
        - httptools:         0.5.0
        - httpx:             0.23.3
        - idna:              3.4
        - importlib-metadata: 6.0.0
        - importlib-resources: 5.12.0
        - inquirer:          3.1.3
        - ipykernel:         6.22.0
        - ipython:           8.12.0
        - ipython-genutils:  0.2.0
        - ipywidgets:        8.0.6
        - isoduration:       20.11.0
        - itsdangerous:      2.1.2
        - jax:               0.4.6
        - jaxlib:            0.4.6
        - jedi:              0.18.2
        - jinja2:            3.1.2
        - jsonpointer:       2.3
        - jsonschema:        4.17.3
        - jupyter:           1.0.0
        - jupyter-client:    8.1.0
        - jupyter-console:   6.6.3
        - jupyter-core:      5.3.0
        - jupyter-events:    0.6.3
        - jupyter-server:    2.5.0
        - jupyter-server-terminals: 0.4.4
        - jupyterlab-pygments: 0.2.2
        - jupyterlab-widgets: 3.0.7
        - keras:             2.11.0
        - keras-rl:          0.4.2
        - kiwisolver:        1.4.4
        - libclang:          15.0.6.1
        - lightning:         2.0.0
        - lightning-cloud:   0.5.32
        - lightning-utilities: 0.8.0
        - markdown:          3.4.1
        - markdown-it-py:    2.2.0
        - markupsafe:        2.1.2
        - matplotlib:        3.7.1
        - matplotlib-inline: 0.1.6
        - mdurl:             0.1.2
        - mistune:           2.0.5
        - mpmath:            1.3.0
        - multidict:         6.0.4
        - nbclassic:         0.5.5
        - nbclient:          0.7.3
        - nbconvert:         7.3.0
        - nbformat:          5.8.0
        - nest-asyncio:      1.5.6
        - networkx:          3.0
        - notebook:          6.5.3
        - notebook-shim:     0.2.2
        - numpy:             1.24.2
        - oauthlib:          3.2.2
        - opt-einsum:        3.3.0
        - ordered-set:       4.1.0
        - orjson:            3.8.7
        - packaging:         23.0
        - pandas:            1.5.3
        - pandocfilters:     1.5.0
        - parso:             0.8.3
        - pexpect:           4.8.0
        - pickleshare:       0.7.5
        - pillow:            9.4.0
        - pip:               23.0.1
        - pkgutil-resolve-name: 1.3.10
        - platformdirs:      3.2.0
        - poke-env:          0.5.0
        - prometheus-client: 0.16.0
        - prompt-toolkit:    3.0.38
        - protobuf:          3.19.6
        - psutil:            5.9.4
        - ptyprocess:        0.7.0
        - pure-eval:         0.2.2
        - pyasn1:            0.4.8
        - pyasn1-modules:    0.2.8
        - pycparser:         2.21
        - pydantic:          1.10.7
        - pygments:          2.14.0
        - pyjwt:             2.6.0
        - pyparsing:         3.0.9
        - pyrsistent:        0.19.3
        - python-dateutil:   2.8.2
        - python-dotenv:     1.0.0
        - python-editor:     1.0.4
        - python-json-logger: 2.0.7
        - python-multipart:  0.0.6
        - pytorch-lightning: 2.0.0
        - pytz:              2023.2
        - pyyaml:            6.0
        - pyzmq:             25.0.2
        - qtconsole:         5.4.2
        - qtpy:              2.3.1
        - readchar:          4.0.5
        - requests:          2.28.2
        - requests-oauthlib: 1.3.1
        - rfc3339-validator: 0.1.4
        - rfc3986:           1.5.0
        - rfc3986-validator: 0.1.1
        - rich:              13.3.3
        - rl:                3.1
        - rsa:               4.9
        - scipy:             1.10.1
        - seaborn:           0.12.2
        - send2trash:        1.8.0
        - setuptools:        49.2.1
        - six:               1.16.0
        - sniffio:           1.3.0
        - soupsieve:         2.4
        - stack-data:        0.6.2
        - starlette:         0.22.0
        - starsessions:      1.3.0
        - sympy:             1.11.1
        - tabulate:          0.9.0
        - tensorboard:       2.11.2
        - tensorboard-data-server: 0.6.1
        - tensorboard-plugin-wit: 1.8.1
        - tensorflow-estimator: 2.11.0
        - tensorflow-hub:    0.13.0
        - tensorflow-macos:  2.11.0
        - tensorflowjs:      4.2.0
        - termcolor:         2.2.0
        - terminado:         0.17.1
        - tf-nightly-2.0-preview: 2.0.1
        - tinycss2:          1.2.1
        - torch:             2.0.0
        - torchmetrics:      0.11.4
        - torchvision:       0.15.1
        - tornado:           6.2
        - tqdm:              4.65.0
        - traitlets:         5.9.0
        - typing-extensions: 4.5.0
        - ujson:             5.7.0
        - uri-template:      1.2.0
        - urllib3:           1.26.15
        - uvicorn:           0.21.1
        - uvloop:            0.17.0
        - watchfiles:        0.18.1
        - wcwidth:           0.2.6
        - webcolors:         1.13
        - webencodings:      0.5.1
        - websocket-client:  1.5.1
        - websockets:        10.4
        - werkzeug:          2.2.3
        - wheel:             0.38.4
        - widgetsnbextension: 4.0.7
        - wrapt:             1.15.0
        - yarl:              1.8.2
        - zipp:              3.15.0
* System:
        - OS:                Darwin
        - architecture:
                - 64bit
                - 
        - processor:         arm
        - python:            3.8.9
        - version:           Darwin Kernel Version 21.2.0: Sun Nov 28 20:28:41 PST 2021; root:xnu-8019.61.5~1/RELEASE_ARM64_T6000

</details>




cc @borda

**Comment by ryan597** [comment]
This is an artifact of you having created a dataset in memory. If you use a practical/realistic method such as loading data from a `Dataset` using the `DataLoader` class then you will see how much slower `num_workers=0` is, especially on CPU because you block the process which is actually running the model.

Further, this will get much, much worse if you actually have to do some processing on your data as you load it in.

Here's an example using some of your code, but I made the model use a conv2d to simulate images.

```python
import random
import time

import lightning.pytorch as pl
import torch
from torch.utils.data import DataLoader, Dataset
from lightning.pytorch.callbacks import TQDMProgressBar, EarlyStopping


class Model(pl.LightningModule):
    def __init__(self):
        super(Model, self).__init__()
        self.conv = torch.nn.Conv2d(3, 3, 3, padding='same')

    def forward(self, x):
        return self.conv(x)

    def training_step(self, train_batch, batch_idx):
        mse_l = torch.nn.MSELoss()
        x, y = train_batch
        logits = self.forward(x)
        loss = mse_l(logits, y)
        self.log('train_loss', loss)
        return loss

    def validation_step(self, val_batch, batch_idx):
        mse_l = torch.nn.MSELoss()
        x, y = val_batch
        logits = self.forward(x)
        loss = mse_l(logits, y)
        self.log('val_loss', loss)
        return loss

    def configure_optimizers(self):
        optimizer = torch.optim.Adam(self.parameters(), lr=1e-3)
        return optimizer


class Data(Dataset):
    def __init__(self, length, do_processing):
        super().__init__()
        self.length = length
        self.do_processing = do_processing
    
    def __len__(self):
        return self.length

    def __getitem__(self, index):
        if self.do_processing:
            time.sleep(0.01)
        return [torch.rand(3, 256, 256), torch.rand(3, 256, 256)]

def test(num_workers, do_processing):
    train_data = Data(3200, do_processing)
    validation_data = Data(320, do_processing)

    train_loader = DataLoader(train_data, batch_size=32, num_workers=num_workers)
    validation_loader = DataLoader(validation_data, batch_size=32, num_workers=num_workers)

    trainer = pl.Trainer(max_epochs=1,
                        accelerator='cpu',
                        devices=1,
                        callbacks=[EarlyStopping(monitor="val_loss", mode="min", patience=3),
                                    TQDMProgressBar(refresh_rate=1)],
                        log_every_n_steps=10
                        )

    model = Model()

    start_time = time.time()
    trainer.fit(model=model, train_dataloaders=train_loader, val_dataloaders=validation_loader)
    end_time = time.time()
    print(f"num_workers= {num_workers}\tdo processing= {do_processing}")
    print(f"training took = {end_time -start_time}s\n")

if __name__ == "__main__":
    test(0, False)
    test(15, False)
    test(0, True)
    test(15, True)
```

With this simple case for 1 epoch I got

```
num_workers= 0	do processing= False
training took = 6.58838677406311s

num_workers= 15	do processing= False
training took = 5.423422574996948s

num_workers= 0	do processing= True
training took = 48.26198148727417s

num_workers= 15	do processing= True
training took = 7.059680938720703s
```
So you can see the huge difference that having `num_workers=0` makes when you have to do even a tiny amount of processing of the data. This processing could be opening files, decoding images, normalizing, getting tokens in text etc.

If you're curious, here's the same tests on a GPU.

```
num_workers= 0	do processing= False
training took = 4.558110475540161s

num_workers= 15	do processing= False
training took = 3.4102823734283447s

num_workers= 0	do processing= True
training took = 46.69817566871643s

num_workers= 15	do processing= True
training took = 5.927751302719116s
```
